torch/export/_decomp_utils.py

```diff
@@ -0,0 +1,178 @@
+import itertools
+from typing import Callable, Dict
+
+import torch
+from torch._export.utils import (
+    _collect_all_valid_cia_ops,
+    _collect_all_valid_cia_ops_for_aten_namespace,
+    _get_decomp_for_cia,
+    _is_aten_op,
+    _is_custom_op,
+    _is_preservable_cia_op,
+)
+
+
+class CustomDecompTable(Dict[torch._ops.OperatorBase, Callable]):
+    """
+    This is a custom dictionary that is specifically used for handling decomp_table in export.
+    The reason we need this is because in the new world, you can only *delete* an op from decomp
+    table to preserve it. This is problematic for custom ops because we don't know when the custom
+    op will actually be loaded to the dispatcher. As a result, we need to record the custom ops operations
+    until we really need to materialize it (which is when we run decomposition pass.)
+
+    Invariants we hold are:
+    1. All aten decomp is loaded at the init time
+    2. We materialize ALL ops when user ever reads from the table to make it more likely
+       that dispatcher picks up the custom op.
+    3. If it is write operation, we don't necessarily materialize
+    4. We load the final time during export, right before calling run_decompositions()
+    """
+
+    def __init__(self):
+        super().__init__()
+        from torch._decomp import _core_aten_decompositions_post_autograd
+
+        # For aten ops, we load them up in the beginning
+        self.aten_decomp_table = _core_aten_decompositions_post_autograd()
+
+        for op in _collect_all_valid_cia_ops_for_aten_namespace():
+            self.aten_decomp_table[op] = _get_decomp_for_cia(op)
+
+        # This is to track the *pending* deleted custom ops that haven't been materialized yet
+        self.deleted_custom_ops = set()
+        # When we can materialize a custom op, we use this table
+        self.additional_custom_op_decomp = {}
+        # When this is true, there shouldn't be any pending operations in the table.
+        self.has_materialized = False
+
+    def __getitem__(self, key):
+        self._materialize_if_needed()
+
+        if _is_aten_op(key):
+            return self.aten_decomp_table[key]
+
+        assert _is_custom_op(key)
+        return self.additional_custom_op_decomp[key]
+
+    def __setitem__(self, key, value):
+        if _is_aten_op(key):
+            self.aten_decomp_table[key] = value
+            return
+
+        assert _is_custom_op(key)
+
+        self.additional_custom_op_decomp[key] = value
+
+        if key in self.deleted_custom_ops:
+            self.deleted_custom_ops.remove(key)
+
+    def keys(self):
+        self._materialize_if_needed()
+
+        return itertools.chain(
+            self.aten_decomp_table.keys(), self.additional_custom_op_decomp.keys()
+        )
+
+    def __delitem__(self, key):
+        self.pop(key)
```

Comment:
hmm is there any reason that we can't just inherit the existing impl from dict, since we are subclassing dict here? (`__delitem__` and `__missing__` don't seem like they need any special handling)

Reply:
Delitem needs to be special because we could have pending operators in self.deleted_custom_ops set.